File: tools/ios-crawler/src/frames.js
/**
 * Frame extraction, with two backends.
 *
 *   ffmpeg          — used when it is on PATH. Faster, and reads more formats.
 *   motvin-frames   — AVFoundation, compiled on demand from frames.m.
 *
 * The fallback is not a nicety. Installing ffmpeg means writing into the
 * Homebrew prefix, which on a managed Mac belongs to an admin account the
 * person running this does not have. AVFoundation and clang are both already
 * present — clang because the Command Line Tools are required anyway — so the
 * fallback needs no permission from anyone.
 *
 * Both backends emit frame-00001.png … in chronological order, and alongside
 * every frame a tiny raw RGB thumbnail (THUMB.width × THUMB.height, 8-bit, no
 * header). The thumbnails are what the segmenter works on: deciding whether the
 * UI held still, scrolled, opened a sheet or flashed a loading state takes a
 * few thousand pixels, and producing those while the frame is already decoded
 * is what keeps a long recording quick to analyse.
 */

import { existsSync, mkdirSync, readdirSync, readFileSync, statSync } from 'node:fs';
import { dirname, join } from 'node:path';
import { fileURLToPath } from 'node:url';
import { has, run } from './exec.js';
import { log } from './log.js';
import { THUMB } from './hash.js';

const HERE = dirname(fileURLToPath(import.meta.url));
const SOURCE = join(HERE, 'frames.m');
const BIN_DIR = join(HERE, '..', '.bin');
const BINARY = join(BIN_DIR, 'motvin-frames');

const FRAMEWORKS = [
  'Foundation',
  'AVFoundation',
  'CoreMedia',
  'CoreGraphics',
  'ImageIO',
  'CoreServices',
];

/**
 * Compiles the helper if it is missing or older than its source. Cheap enough
 * to check every run; the compile itself takes about a second and happens once.
 */
export async function buildFrameExtractor() {
  if (existsSync(BINARY) && statSync(BINARY).mtimeMs >= statSync(SOURCE).mtimeMs) {
    return BINARY;
  }

  if (!(await has('clang'))) {
    throw new Error(
      'clang is not available, so the built-in frame reader cannot be built.\n' +
        '  Install the Command Line Tools: xcode-select --install\n' +
        '  Or install ffmpeg instead: brew install ffmpeg',
    );
  }

  mkdirSync(BIN_DIR, { recursive: true });
  log.detail('building the frame reader (once)…');

  const result = await run(
    'clang',
    [
      '-fobjc-arc',
      '-O2',
      // kUTTypePNG is formally deprecated in favour of UniformTypeIdentifiers,
      // but still the portable spelling across the SDK versions this has to
      // build against. The warning is noise, not a signal.
      '-Wno-deprecated-declarations',
      ...FRAMEWORKS.flatMap((framework) => ['-framework', framework]),
      '-o',
      BINARY,
      SOURCE,
    ],
    { timeout: 120_000 },
  );

  if (result.failed) {
    throw new Error(`could not build the frame reader: ${result.stderr.trim().split('\n').slice(0, 3).join(' ')}`);
  }
  return BINARY;
}

/**
 * Writes frames from `videoPath` into `framesDir` at `fps` per second.
 *
 * @returns {Promise<{paths: string[], thumbs: Buffer[], fps: number, backend: string}>}
 *   `thumbs[i]` is the raw RGB thumbnail of `paths[i]`.
 */
export async function extractFrames(videoPath, framesDir, fps) {
  mkdirSync(framesDir, { recursive: true });

  const backend = (await has('ffmpeg')) ? 'ffmpeg' : 'avfoundation';
  if (backend === 'ffmpeg') {
    // Two outputs from one decode: the frames, and every frame's thumbnail
    // concatenated into a single raw file that is split below.
    const result = await run(
      'ffmpeg',
      [
        '-nostdin',
        '-loglevel',
        'error',
        '-i',
        videoPath,
        '-vf',
        `fps=${fps}`,
        join(framesDir, 'frame-%05d.png'),
        '-vf',
        `fps=${fps},scale=${THUMB.width}:${THUMB.height}:flags=area,format=rgb24`,
        '-f',
        'rawvideo',
        join(framesDir, 'thumbs.rgb'),
      ],
      { timeout: 900_000 },
    );
    if (result.failed) {
      const reason = result.stderr.trim().split('\n').slice(-2).join(' ') || `exit ${result.code}`;
      throw new Error(`ffmpeg could not read the video: ${reason}`);
    }
  } else {
    const binary = await buildFrameExtractor();
    const result = await run(binary, [videoPath, framesDir, String(fps), String(THUMB.width), String(THUMB.height)], {
      timeout: 900_000,
    });
    if (result.failed) {
      throw new Error(result.stderr.trim().split('\n')[0] || `frame reader exited ${result.code}`);
    }
  }

  const paths = readdirSync(framesDir)
    .filter((file) => /^frame-\d+\.png$/.test(file))
    .sort()
    .map((file) => join(framesDir, file));

  if (!paths.length) throw new Error('no frames came out of the video');

  const thumbs = readThumbs(framesDir, paths, backend);
  return { paths, thumbs, fps, backend };
}

/**
 * Pairs each frame with its thumbnail.
 *
 * The AVFoundation reader writes one `.rgb` beside each frame; ffmpeg writes
 * them all into `thumbs.rgb` in frame order. Either way the result is one
 * buffer of THUMB.width × THUMB.height × 3 bytes per frame. A thumbnail that
 * cannot be found is an error rather than a gap: the segmenter compares
 * neighbours, and a missing neighbour would silently distort the timeline.
 */
function readThumbs(framesDir, paths, backend) {
  const size = THUMB.width * THUMB.height * 3;

  if (backend === 'ffmpeg') {
    const all = readFileSync(join(framesDir, 'thumbs.rgb'));
    const count = Math.floor(all.length / size);
    if (count < paths.length) {
      throw new Error(`ffmpeg produced ${paths.length} frames but only ${count} thumbnails`);
    }
    return paths.map((_, index) => all.subarray(index * size, (index + 1) * size));
  }

  return paths.map((path) => {
    const thumbPath = path.replace(/\.png$/, '.rgb');
    if (!existsSync(thumbPath)) throw new Error(`thumbnail missing for ${path}`);
    const thumb = readFileSync(thumbPath);
    if (thumb.length !== size) throw new Error(`thumbnail for ${path} is ${thumb.length} bytes, expected ${size}`);
    return thumb;
  });
}

File: tools/ios-crawler/src/frames.m
/*
 * motvin-frames — pull frames out of a video using only what macOS already has.
 *
 *   motvin-frames <video> <output-dir> <fps> [thumb-width thumb-height]
 *
 * ffmpeg is the obvious tool for this and the crawler prefers it when present.
 * But installing it needs Homebrew, and on a managed Mac the Homebrew prefix
 * belongs to an admin account the person running this tool is not. Rather than
 * making video ingest depend on someone else's permission, this does the same
 * job with AVFoundation, which ships with the OS, and clang, which ships with
 * the Command Line Tools that are already required.
 *
 * Built on demand by src/frames.js and cached in .bin/.
 *
 * Two files per frame:
 *
 *   frame-00001.png   the full frame, for the library
 *   frame-00001.rgb   a tiny RGB thumbnail (default 32×64, 8-bit, no header)
 *
 * The thumbnail is what the segmenter reads. Every decision about whether the
 * UI held still, scrolled, opened a sheet or flashed a loading state is made
 * on a few thousand pixels, so producing them here — while the decoded frame
 * is already in memory — is what keeps a three-minute recording from taking
 * minutes to analyse. Drawn with area-averaging, so a flat splash and a busy
 * feed both reduce honestly.
 *
 * Frames are numbered from 1 so the caller can sort them by name and get
 * chronological order, matching ffmpeg's output convention exactly.
 */

#import <AVFoundation/AVFoundation.h>
#import <CoreGraphics/CoreGraphics.h>
#import <Foundation/Foundation.h>
#import <ImageIO/ImageIO.h>

static BOOL WritePng(CGImageRef image, NSString *path) {
  CFURLRef url = (__bridge CFURLRef)[NSURL fileURLWithPath:path];
  CGImageDestinationRef destination = CGImageDestinationCreateWithURL(url, kUTTypePNG, 1, NULL);
  if (!destination) return NO;
  CGImageDestinationAddImage(destination, image, NULL);
  BOOL ok = CGImageDestinationFinalize(destination);
  CFRelease(destination);
  return ok;
}

/** Draws the frame into a small RGB bitmap and writes the raw bytes. */
static BOOL WriteThumb(CGImageRef image, NSString *path, int width, int height) {
  size_t bytesPerRow = (size_t)width * 4;
  unsigned char *pixels = calloc(bytesPerRow * height, 1);
  if (!pixels) return NO;

  CGColorSpaceRef space = CGColorSpaceCreateDeviceRGB();
  CGContextRef context = CGBitmapContextCreate(
      pixels, width, height, 8, bytesPerRow, space,
      kCGImageAlphaNoneSkipLast | kCGBitmapByteOrder32Big);
  CGColorSpaceRelease(space);
  if (!context) {
    free(pixels);
    return NO;
  }

  CGContextSetInterpolationQuality(context, kCGInterpolationHigh);
  CGContextDrawImage(context, CGRectMake(0, 0, width, height), image);
  CGContextRelease(context);

  // Pack to RGB, top row first — CoreGraphics draws bottom-up.
  NSMutableData *out = [NSMutableData dataWithLength:(NSUInteger)width * height * 3];
  unsigned char *dst = out.mutableBytes;
  for (int y = 0; y < height; y++) {
    const unsigned char *row = pixels + (size_t)(height - 1 - y) * bytesPerRow;
    for (int x = 0; x < width; x++) {
      *dst++ = row[x * 4];
      *dst++ = row[x * 4 + 1];
      *dst++ = row[x * 4 + 2];
    }
  }
  free(pixels);
  return [out writeToFile:path atomically:YES];
}

int main(int argc, const char **argv) {
  @autoreleasepool {
    if (argc < 4) {
      fprintf(stderr, "usage: motvin-frames <video> <output-dir> <fps> [thumb-width thumb-height]\n");
      return 2;
    }

    NSString *videoPath = [NSString stringWithUTF8String:argv[1]];
    NSString *outputDir = [NSString stringWithUTF8String:argv[2]];
    double fps = atof(argv[3]);
    if (fps <= 0) fps = 2;
    int thumbWidth = argc > 5 ? atoi(argv[4]) : 32;
    int thumbHeight = argc > 5 ? atoi(argv[5]) : 64;
    if (thumbWidth <= 0 || thumbHeight <= 0) {
      thumbWidth = 32;
      thumbHeight = 64;
    }

    NSURL *url = [NSURL fileURLWithPath:videoPath];
    AVURLAsset *asset = [AVURLAsset URLAssetWithURL:url options:nil];

    if ([[asset tracksWithMediaType:AVMediaTypeVideo] count] == 0) {
      fprintf(stderr, "no video track in %s\n", argv[1]);
      return 1;
    }

    double duration = CMTimeGetSeconds(asset.duration);
    if (!(duration > 0)) {
      fprintf(stderr, "could not read the duration of %s\n", argv[1]);
      return 1;
    }

    AVAssetImageGenerator *generator = [[AVAssetImageGenerator alloc] initWithAsset:asset];
    generator.appliesPreferredTrackTransform = YES;
    // A little under half a sampling step either way. Exact seeking forces a
    // decode from the previous keyframe for every single frame, which turns a
    // three-minute recording into a several-minute wait for no benefit — this
    // is sampling a screen recording, not cutting a film. The tolerance is tied
    // to the rate so that two requested frames can never resolve to the same
    // decoded one.
    double tolerance = fmin(0.1, 0.4 / fps);
    generator.requestedTimeToleranceBefore = CMTimeMakeWithSeconds(tolerance, 600);
    generator.requestedTimeToleranceAfter = CMTimeMakeWithSeconds(tolerance, 600);

    NSFileManager *files = [NSFileManager defaultManager];
    [files createDirectoryAtPath:outputDir withIntermediateDirectories:YES attributes:nil error:NULL];

    int written = 0;
    int failed = 0;
    double step = 1.0 / fps;

    for (double t = 0; t < duration; t += step) {
      @autoreleasepool {
        CMTime time = CMTimeMakeWithSeconds(t, 600);
        NSError *error = nil;
        CGImageRef image = [generator copyCGImageAtTime:time actualTime:NULL error:&error];
        if (!image) {
          // One unreadable frame is not worth abandoning the recording for; a
          // run of them will show up as a low frame count in the caller.
          failed++;
          continue;
        }
        NSString *name = [NSString stringWithFormat:@"frame-%05d", written + 1];
        NSString *base = [outputDir stringByAppendingPathComponent:name];
        BOOL ok = WritePng(image, [base stringByAppendingPathExtension:@"png"]);
        if (ok) ok = WriteThumb(image, [base stringByAppendingPathExtension:@"rgb"], thumbWidth, thumbHeight);
        CGImageRelease(image);
        if (ok) {
          written++;
        } else {
          failed++;
        }
      }
    }

    if (written == 0) {
      fprintf(stderr, "no frames could be read from %s\n", argv[1]);
      return 1;
    }

    fprintf(stdout, "%d\n", written);
    if (failed > 0) fprintf(stderr, "skipped %d unreadable frame(s)\n", failed);
    return 0;
  }
}
